sklearn/decomposition/online_lda.py

```diff
@@ -0,0 +1,668 @@
+            Document topic distribution. In the literature, this is called `gamma`.
+
+        sub_sampling : boolean, optional, (default=False)
+            Compensate for subsampling of documents.
+            It is used in calcuate bound in online learning.
+
+        Returns
+        -------
+        score : float
+
+        """
+
+        n_samples, n_topics = doc_topic_distr.shape
+        score = 0
+        dirichlet_doc_topic = _dirichlet_expectation(doc_topic_distr)
+
+        X_data = X.data
+        X_indices = X.indices
+        X_indptr = X.indptr
+
+        # E[log p(docs | theta, beta)]
+        for d in xrange(0, n_samples):
+            ids = X_indices[X_indptr[d]:X_indptr[d + 1]]
+            cnts = X_data[X_indptr[d]:X_indptr[d + 1]]
+            id_length = len(ids)
+            norm_phi = np.zeros(id_length)
+            for i in xrange(0, id_length):
+                temp = dirichlet_doc_topic[d, :] + self.dirichlet_component_[:, ids[i]]
+                tmax = temp.max()
+                norm_phi[i] = np.log(np.sum(np.exp(temp - tmax))) + tmax
+            score += np.sum(cnts * norm_phi)
+
+        # compute E[log p(theta | alpha) - log q(theta | gamma)]
+        score += np.sum((self.doc_topic_prior_ - doc_topic_distr) * dirichlet_doc_topic)
+        score += np.sum(gammaln(doc_topic_distr) - gammaln(self.doc_topic_prior_))
+        score += np.sum(
+            gammaln(self.doc_topic_prior_ * self.n_topics) - gammaln(np.sum(doc_topic_distr, 1)))
+
+        # Compensate for the subsampling of the population of documents
+        # E[log p(beta | eta) - log q (beta | lambda)]
+        score += np.sum((self.topic_word_prior_ - self.components_) * self.dirichlet_component_)
+        score += np.sum(gammaln(self.components_) - gammaln(self.topic_word_prior_))
+        score += np.sum(gammaln(self.topic_word_prior_ * self.n_features)
+                        - gammaln(np.sum(self.components_, 1)))
+
+        # Compensate for the subsampling of the population of documents
+        if sub_sampling:
+            doc_ratio = float(self.n_samples) / n_samples
+            score *= doc_ratio
+
+        return score
+
+    def score(self, X, y=None):
+        """
+        use approximate log-likelihood as score
+
+        Parameters
+        ----------
+        X : sparse matrix, shape=(n_samples, n_features)
+             Document word matrix.
+
+        Returns
+        -------
+        score : float
+            Use approximate bound as score.
+        """
+
+        X = self._to_csr(X, "LatentDirichletAllocation.score")
+
+        doc_topic_distr = self.transform(X)
+        score = self._approx_bound(X, doc_topic_distr, sub_sampling=False)
+        return score
+
+    def perplexity(self, X, doc_topic_distr, sub_sampling=False):
+        """
+        calculate approximate perplexity for data X and topic distribution `gamma`.
+        Perplexity is defined as exp(-1. * log-likelihood per word)
```

Comment:
For perplexity, the definition and calculation is from the [literature](https://www.cs.princeton.edu/~blei/papers/HoffmanBleiBach2010b.pdf). (formula 15 & 16)

[here](http://stats.stackexchange.com/questions/18167/how-to-calculate-perplexity-of-a-holdout-with-latent-dirichlet-allocation) is some discussion on stackexchange .